src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -163,37 +163,57 @@ public NodeStateTable<T> CreateCachedTableWithUpdatedSteps<TInput>(NodeStateTabl
 
         public sealed class Builder
         {
-            private readonly ArrayBuilder<TableEntry> _states;
+            /// <summary>
+            /// Only mutate this field from within <see cref="AddState(TableEntry)"/>.
+            /// </summary>
+            private readonly ArrayBuilder<TableEntry> _states_doNotMutateDirectly;
             private readonly NodeStateTable<T> _previous;
 
+            private readonly IEqualityComparer<T> _equalityComparer;
+
             private readonly string? _name;
             private readonly ArrayBuilder<IncrementalGeneratorRunStep>? _steps;
 
             [MemberNotNullWhen(true, nameof(_steps))]
             public bool TrackIncrementalSteps => _steps is not null;
 
-            internal Builder(NodeStateTable<T> previous, string? name, bool stepTrackingEnabled)
+            private bool _unchangedFromPrevious = true;
+
+            internal Builder(NodeStateTable<T> previous, string? name, bool stepTrackingEnabled, IEqualityComparer<T>? equalityComparer)
             {
-                _states = ArrayBuilder<TableEntry>.GetInstance();
+                _states_doNotMutateDirectly = ArrayBuilder<TableEntry>.GetInstance(previous.Count);
                 _previous = previous;
                 _name = name;
+                _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
                 if (stepTrackingEnabled)
                 {
                     _steps = ArrayBuilder<IncrementalGeneratorRunStep>.GetInstance();
                 }
             }
 
+            private void AddState(TableEntry entry)
+            {
+                var currentindex = _states_doNotMutateDirectly.Count;
+                _states_doNotMutateDirectly.Add(entry);
+
+                if (_unchangedFromPrevious && currentindex < _previous._states.Length)
+                {
+                    var previousEntry = _previous._states[currentindex];
+                    _unchangedFromPrevious = entry.Matches(previousEntry, _equalityComparer);
+                }
+            }
```

Comment:
i coudl also do this as a final pass over the states in ToImmutableAndFree.  But it seemed nice to be able to just do the check as we're adding the items.  It would be safer though to do as a final pass as we wouldn't have watchdog items getting added to our own builder.